File: src/top_layer.js
/* Why the engine was painting an element from the top layer, or nothing if it was not.

   The top layer is a user-agent list, not a property of any element, and nothing in the
   document declares membership of it. Three calls put an element there and they are not
   variants of one thing: they differ in what they do to the rest of the page, they are
   selected by three different pseudo-classes, and — the reason this is a reason and not a
   flag — a recreation puts an element back by making the same call the page made, so the
   record has to say which call that was.

   Recording inertness instead is what loses the largest member. `:modal` is specified as
   excluding all interaction with everything outside the element, which is true of a dialog
   opened with `showModal()` and of the fullscreen element and deliberately false of a
   popover, whose whole purpose is that the page behind it stays live. So the top layer is
   strictly wider than `:modal`, and an open popover answering that question records `false`
   — not "unknown", but a positive claim that it was an ordinary in-flow box.

   Ordered by how specific the answer is rather than by preference. `:modal` matches the
   fullscreen element as well as a modal dialog, so fullscreen is asked first or it would be
   reported as something a recreation may call `showModal()` on. A popover is never `:modal`,
   so its position is free; it leads because it is the common case.

   Fullscreen is recorded and never replayed. `requestFullscreen()` requires transient user
   activation, which a recreation loading itself does not have, so there is no call to make —
   but the record still has to distinguish it, or a later stage reading "in the top layer"
   would replay it as a dialog. Naming the reason is what lets one field carry a promotion
   with no replay beside two with different ones. */
const TOP_LAYER_REASONS = [
  [':popover-open', 'popover'],
  [':fullscreen', 'fullscreen'],
  [':modal', 'modal']
];
const recreateTopLayer = element => {
  for (const [selector, reason] of TOP_LAYER_REASONS) {
    try {
      if (element.matches(selector)) return reason;
    } catch (unsupported) {
      /* An engine that does not know a selector cannot have promoted anything for that
         reason, so the remaining ones are still worth asking. Throwing here would cost the
         page every promotion, including the ones the engine does support. */
    }
  }
  return '';
};

File: src/top_layer.rs
//! The one place that decides why an element was in the top layer.
//!
//! Both node-record producers and the generated-box gate ask this question, and all three
//! need the same answer: a promotion the capture does not record cannot be replayed, and a
//! `::backdrop` box is generated for a top-layer element whatever put it there, so a gate
//! that asks a narrower question declines to look for a box the engine really generated.
//! Rendered into the one bundle every reader already renders, so the three cannot drift.

pub const SOURCE: &str = include_str!("top_layer.js");

/// The promotion a node record carries, and what the recreation does about it.
///
/// A string rather than a flag because the three ways into the top layer replay as three
/// different calls, and an empty string because that is what the engine's own answer reduces
/// to for an element the page never promoted.
#[derive(Debug, Default, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct Promotion(pub String);

impl Promotion {
    /// The calls that put an element back in the top layer, in the order a recreation has to
    /// make them.
    ///
    /// Ordered by dismissal, which is a fact about the platform rather than a preference:
    /// showing a dialog modally hides every auto popover in the document, while showing a
    /// popover hides no dialog. Replaying in document order therefore deletes whichever
    /// popover happens to precede a modal — and a popover the recreation opened and then
    /// closed again is indistinguishable from the one it never opened. Making the dismissing
    /// call first leaves it nothing to dismiss.
    ///
    /// Fullscreen is absent because it has no call, which is the same reason it carries no
    /// marker: `requestFullscreen()` needs transient user activation a recreation rendering
    /// itself does not have.
    pub const REPLAY: [(&'static str, &'static str); 2] =
        [("modal", "showModal"), ("popover", "showPopover")];

    /// Whether the page had this element in the top layer at all.
    pub fn promoted(&self) -> bool {
        !self.0.is_empty()
    }

    /// The serde companion to [`Promotion::promoted`], so an unpromoted element costs the
    /// record nothing.
    pub fn absent(&self) -> bool {
        !self.promoted()
    }

    /// The method the recreation calls to put the element back, if there is one.
    ///
    /// Fullscreen has none: `requestFullscreen()` needs transient user activation, which a
    /// recreation rendering itself does not have. It is still recorded, because a reader that
    /// cannot see the promotion cannot see that it was lost either.
    pub fn replay(&self) -> Option<&'static str> {
        Self::REPLAY
            .iter()
            .find(|(reason, _)| *reason == self.0)
            .map(|(_, call)| *call)
    }
}

#[cfg(test)]
#[path = "top_layer_tests.rs"]
mod tests;
